### ISSUE TITLE
nrow_subset_linter() should understand data.table '.N' usage too

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2315]
```r
DT[conds, .(.N)]

# and
nrow(DT[(conds)])

# but NOT
DT[conds, .N]

# or
nrow(DT[conds])

# or
DT[conds, .(.N), by = by]

# or
nrow(DT[conds, j, by = by])
```

Generally I think this one is a lot tougher to get right statically since how  `[` operates can be hard to tell statically.

E.g. `nrow(DT[conds])` should be a join if `cond` is a symbol, which _might_ be better to calculate some other way, though "count the rows after a join" is hard to describe without "doing the join".

But there are definitely some rules we can incorporate here.

**Comment by MichaelChirico** [comment]
Copying test cases out from the script:

```r
test_that("nrow_subset_linter skips allowed usages", {
  # nrow can be avoided here by using uniqueN(),
  #   but it's not easy to tell statically whether uniqueN() applies
  lintr::expect_lint(
    "nrow(DT[x == y, 1, by = grp])",
    NULL,
    nrow_subset_linter
  )
})

lintr::expect_lint(
  "x %>% subset(y == z) %>% nrow()",
  "Use arithmetic to count the number of rows satisfying a condition",
  nrow_subset_linter
)
test_that("nrow_subset_linter blocks [ cases", {
  # data.frame subsetting (NB: replacement doesn't use na.rm = TRUE)
  lintr::expect_lint(
    "nrow(x[x$y == x$z, ])",
    "Use arithmetic to count the number of rows satisfying a condition",
    nrow_subset_linter
  )

  # data.table subsetting (NB: replacement needs na.rm = TRUE)
  lintr::expect_lint(
    "x[y == z, ]",
    "Use arithmetic to count the number of rows satisfying a condition",
    nrow_subset_linter
  )
})

test_that("nrow_subset_linter blocks dplyr::filter() cases", {
  lintr::expect_lint(
    "x %>% filter(y == z) %>% nrow()",
    "Use arithmetic to count the number of rows satisfying a condition",
    nrow_subset_linter
  )

  lintr::expect_lint(
    "nrow(dplyr::filter(x, y == z))",
    "Use arithmetic to count the number of rows satisfying a condition",
    nrow_subset_linter
  )
})
```

**Comment by MichaelChirico** [comment]
Update this not to be {data.table}-specific. If anything this is about ``nrow(`[`(x, ...))`` cases where we can tell statically that `[` can be avoided. It's not an easy question and I'm skeptical we'll get anything too good here. Bumping from 3.2.0.
